### PR TITLE
feat(ui): readiness gates failures alert in pod summary tab #12515

### LINKED ISSUE #12515: Note readiness gate failures in Pod summary UI
# Summary

The Pod summary should show some indication that a readiness gate is failing.

# Motivation

Readiness probes produce events when they fail. Users know to look at the Events tab for those failures. But for readiness gates, you have to dig through the `status` field to figure out what's wrong.

# Proposal

Show a small alert near the top of the Pod summary UI indicating that a readiness gate is failing.

### PR DESCRIPTION
**Pull request by Marvin9**
fixes #12515

readiness_gates_1
readiness_gates_2

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
This looks awesome. @rbreeze @keithchong thoughts?

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx [hunk 1]
```diff
@@ -167,8 +166,46 @@ export const ApplicationNodeInfo = (props: {
         });
     }
 
+    const getPodReadinessGatesState = React.useCallback(_getPodReadinessGatesState, [props.live]);
+
+    let ReadinessGatesFailedWarning: React.ReactNode = null;
+    if (props.live && props.node.kind === 'Pod') {
+        const readinessGatesState = getPodReadinessGatesState(props.live);
+
+        if (readinessGatesState.failedConditions.length > 0 || readinessGatesState.nonExistConditions.length > 0) {
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-node-info/application-node-info.tsx:175]: These should have null protection, i.e. `(readinessGatesState?.failedConditions || []).length`

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx [hunk 2]
```diff
@@ -167,8 +166,46 @@ export const ApplicationNodeInfo = (props: {
...
+            ReadinessGatesFailedWarning = (
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-node-info/application-node-info.tsx:176]: IMO the code would read better if this were a separate component which received `readinessGatesState` as a prop, and was conditionally rendered inline based on the condition above. LMK if there's a specific reason why you went with this approach instead, either is okay 🙂

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx [hunk 3]
```diff
@@ -167,8 +166,46 @@ export const ApplicationNodeInfo = (props: {
...
+                <div className='white-box white-box__readiness-gates-alert'>
+                    <h5>Readiness Gates Failing: </h5>
+                    <ul>
+                        {readinessGatesState.failedConditions.length > 0 && (
+                            <li>
+                                the status of pod readiness gate{selectPostfix(readinessGatesState.failedConditions, '', 's')}{' '}
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-node-info/application-node-info.tsx:182]: Nitpick: `the` should be capitalized -> `The`

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx [hunk 4]
```diff
@@ -167,8 +166,46 @@ export const ApplicationNodeInfo = (props: {
...
+                                {readinessGatesState.failedConditions
+                                    .map(t => `"${t}"`)
+                                    .join(', ')
+                                    .trim()}{' '}
+                                {selectPostfix(readinessGatesState.failedConditions, 'is', 'are')} False.
+                            </li>
+                        )}
+                        {readinessGatesState.nonExistConditions.length > 0 && (
+                            <li>
+                                corresponding condition{selectPostfix(readinessGatesState.nonExistConditions, '', 's')} of pod readiness gate{' '}
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-node-info/application-node-info.tsx:192]: nit: `corresponding` -> `Corresponding`

File: ui/src/app/applications/components/utils.tsx [hunk 5]
```diff
@@ -945,6 +945,53 @@ export function getPodStateReason(pod: appModels.State): {message: string; reaso
...
 
+export const getPodReadinessGatesState = (pod: appModels.State): {nonExistConditions: string[]; failedConditions: string[]} => {
+    if (!pod.spec?.readinessGates?.length) {
+        return {
+            nonExistConditions: [],
+            failedConditions: []
+        };
+    }
+
+    const existentConditions = new Map<string, boolean>();
```
Comment (rbreeze) [inline ui/src/app/applications/components/utils.tsx:956]: nit: `existent` -> `existing`

File: ui/src/app/applications/components/utils.tsx [hunk 6]
```diff
@@ -945,6 +945,53 @@ export function getPodStateReason(pod: appModels.State): {message: string; reaso
     return {reason, message};
 }
 
+export const getPodReadinessGatesState = (pod: appModels.State): {nonExistConditions: string[]; failedConditions: string[]} => {
```
Comment (rbreeze) [inline ui/src/app/applications/components/utils.tsx:948]: nit: `nonExist` -> `nonExisting`

**Review by rbreeze — CHANGES_REQUESTED** [review]
Great improvement, looks great! Just a few small comments

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx (hunk 1 above)
Comment (Marvin9) [reply ui/src/app/applications/components/application-node-info/application-node-info.tsx:175]: the return of function `getPodReadinessGatesState` is surely an object with given keys and empty array as default

File: ui/src/app/applications/components/application-node-info/application-node-info.tsx (hunk 2 above)
Comment (Marvin9) [reply ui/src/app/applications/components/application-node-info/application-node-info.tsx:176]: Agreed. No specific reason, will move it to separate component

**Review by rbreeze — APPROVED** [review]
LGTM

File: ui/src/app/applications/components/utils.tsx [hunk 7]
```diff
@@ -945,6 +945,53 @@ export function getPodStateReason(pod: appModels.State): {message: string; reaso
...
+
+    const readinessGates: appModels.ReadinessGate[] = pod.spec?.readinessGates || [];
+
+    for (const readinessGate of readinessGates) {
+        if (!existingConditions.has(readinessGate.conditionType)) {
+            nonExistingConditions.push(readinessGate.conditionType);
+        } else if (podConditions.get(readinessGate.conditionType) === false) {
+            failedConditions.push(readinessGate.conditionType);
+        }
+    }
```
Comment (crenshaw-dev) [inline ui/src/app/applications/components/utils.tsx:987]: Leaving this here for my own future reference:

Looks like a readiness gate may have either a corresponding “success” condition (meaning things are fine), a “failed” condition (meaning things are not fine), or no condition at all (meaning things are not yet fine).

**Pull request merged**